File: src/main/java/gruvexp/bbminigames/twtClassic/hazard/Hazard.java
package gruvexp.bbminigames.twtClassic.hazard;

import gruvexp.bbminigames.twtClassic.BotBowsPlayer;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

public abstract class Hazard {

    private HazardChance hazardChance = getDefaultChance();
    private boolean isActive = false;

    public HazardChance getChance() {return hazardChance;}

    public void setChance(HazardChance chance) {hazardChance = chance;}

    public Map<BotBowsPlayer, BukkitRunnable> hazardTimers = new HashMap<>();

    public void triggerOnChance(Collection<BotBowsPlayer> players) {
        if (hazardChance.occurs()) {
            isActive = true;
            announce(players);
            trigger(players);
        }
    }

    public boolean isActive() {
        return isActive;
    }

    public abstract void init(Collection<BotBowsPlayer> players);
    public abstract HazardChance getDefaultChance();

    protected abstract void trigger(Collection<BotBowsPlayer> players); // hazarden starter
    protected abstract HazardMessage getAnnounceMessage();
    private void announce(Collection<BotBowsPlayer> players) {
        HazardMessage msg = getAnnounceMessage();
        players.forEach(p -> {
            p.avatar.message(Component.text(msg.chatHeader, NamedTextColor.DARK_RED)
                    .append(Component.text(" " + msg.chatDescription, NamedTextColor.RED)));
            p.avatar.showTitle(Component.text(msg.screenTitle, NamedTextColor.RED), 4);
        });
    }
    public abstract String getName();

    public abstract Component[] getDescription();

    public abstract String getActionDescription();

    public void end() {
        for (BukkitRunnable timer : hazardTimers.values()) { // stopp timerene
            timer.cancel();
        }
        hazardTimers.clear();
        isActive = false;
    } // stopper hazarden

    public record HazardMessage(
            String chatHeader,
            String chatDescription,
            String screenTitle
    ) {}
}

File: src/main/java/gruvexp/bbminigames/twtClassic/hazard/Hazard.kt
package gruvexp.bbminigames.twtClassic.hazard

import gruvexp.bbminigames.twtClassic.BotBowsPlayer
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.scheduler.BukkitRunnable

abstract class Hazard protected constructor(val type: HazardType) {
    var chance: HazardChance = type.defaultChance
    var isActive: Boolean = false
        private set

    var hazardTimers = mutableMapOf<BotBowsPlayer, BukkitRunnable>()

    fun triggerOnChance(players: Set<BotBowsPlayer>) {
        if (chance.occurs()) {
            isActive = true
            announce(players)
            trigger(players)
        }
    }

    abstract fun init(players: Set<BotBowsPlayer>)

    protected abstract fun trigger(players: Set<BotBowsPlayer>) // hazarden starter
    protected abstract val announceMessage: HazardMessage

    private fun announce(players: Set<BotBowsPlayer>) {
        val msg = announceMessage
        players.forEach {
            it.avatar.message(
                Component.text(msg.chatHeader, NamedTextColor.DARK_RED)
                    .append(Component.text(" ${msg.chatDescription}", NamedTextColor.RED))
            )
            it.avatar.showTitle(Component.text(msg.screenTitle, NamedTextColor.RED), 4)
        }
    }

    abstract val name: String

    abstract val description: Array<TextComponent>

    abstract val actionDescription: String

    open fun end() {
        hazardTimers.values.forEach { it.cancel() }
        hazardTimers.clear()
        isActive = false
    }

    data class HazardMessage(
        val chatHeader: String,
        val chatDescription: String,
        val screenTitle: String
    )
}
